Port of Java PR #267 - Clipping Path Name / Path Information Extraction

kwhopper opened the pull request:
port of Java PR drewnoakes/metadata-extractor#267

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Knot.cs:
```diff
@@ -0,0 +1,49 @@
+        /// </summary>
+        /// <param name="index">location of point to be added in points</param>
+        /// <param name="point">coordinate value to be added to points</param>
+        public void SetPoint(int index, double point)
```
These get and set methods could be replaces with an indexer property if you liked (not available in Java). Up to you.

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Knot.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <summary>
+        /// Get the type of knot (linked or unlinked)
+        /// </summary>
+        public string Type { get; private set; }
```
```suggestion
        public string Type { get; }
```

drewnoakes commented on MetadataExtractor/Formats/Photoshop/PhotoshopDescriptor.cs:
```diff
@@ -47,7 +49,11 @@ public PhotoshopDescriptor(PhotoshopDirectory directory)
+                case PhotoshopDirectory.TagClippingPathName:
+                    return GetClippingPathNameString(tagType);
                 default:
+                    if (tagType >= 0x07D0 && tagType <= 0x0BB6)
```
What are these two numbers? Can they be defined as constants somewhere?

drewnoakes commented on MetadataExtractor/Formats/Photoshop/PhotoshopReader.cs:
```diff
@@ -133,7 +140,23 @@ public DirectoryList Extract(SequentialReader reader, int length)
                             break;
                         default:
-                            directory.Set(tagType, tagBytes);
+                            if (tagType >= 0x07D0 && tagType <= 0x0BB6)
```
Ditto.

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Subpath.cs:
```diff
@@ -0,0 +1,56 @@
+        /// Gets size of knots list
+        /// </summary>
+        /// <returns>size of knots List</returns>
+        public int Size()
```
This could be a property. Alternatively, remove this `Size` member, make `GetKnots` an `IReadOnlyList<Knot>` and allow callers to use `subPath.Knots.Count`, which would feel more idiomatic in .NET. The name "size" is more Java-esque.

kwhopper replied:
That works, until NET3.5 compat kicks in - then it gets a little weird. I might change the name from GetKnots to Knots and "Size" to "KnotsCount". Otherwise, it has to be the Count() Linq method, which will cause an enumeration. I'll post it and see what you think.

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Subpath.cs:
```diff
@@ -0,0 +1,56 @@
+        {
+        }
+
+        public Subpath(string type)
```
You can merge these two constructors and make the parameter optional, eg:

```suggestion
        public Subpath(string type = "")
```

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Subpath.cs:
```diff
@@ -0,0 +1,56 @@
+            return p_knots;
+        }
+
+        public string Type { get; private set; }
```
```suggestion
        public string Type { get; }
```

drewnoakes commented on MetadataExtractor/Formats/Photoshop/Subpath.cs:
```diff
@@ -0,0 +1,56 @@
+    /// <author>Kevin Mott https://github.com/kwhopper</author>
+    public class Subpath
+    {
+        private List<Knot> p_knots = new List<Knot>();
```
```suggestion
        private readonly List<Knot> p_knots = new List<Knot>();
```

drewnoakes commented on MetadataExtractor/Formats/Photoshop/PhotoshopReader.cs:
```diff
@@ -133,7 +140,23 @@ public DirectoryList Extract(SequentialReader reader, int length)
+                                        tagBytes[i] = (byte)description[i - (tagBytes.Length - description.Length - 1)];
+                                }
+                                PhotoshopDirectory.TagNameMap[0x07CF + clippingPathCount] = "Path Info " + clippingPathCount;
+                                directory.Set(0x07CF + clippingPathCount, tagBytes);
```
Ditto, constants.

drewnoakes commented on MetadataExtractor/Formats/Photoshop/PhotoshopReader.cs:
```diff
@@ -71,16 +72,22 @@ public DirectoryList Extract(SequentialReader reader, int length)
 
                     // Some basic bounds checking
-                    if (descriptionLength + pos > length)
+                    if (descriptionLength < 0 || descriptionLength + pos > length)
```
Doesn't `GetByte` return a `byte`, which cannot be negative?

kwhopper commented:
Tried to accommodate all your suggestions. Thanks

drewnoakes commented on MetadataExtractor/Formats/Photoshop/PhotoshopReader.cs:
```diff
@@ -71,16 +72,22 @@ public DirectoryList Extract(SequentialReader reader, int length)
 
                     // A variable number of bytes holding a pascal string (two leading bytes for length).
-                    var descriptionLength = reader.GetByte();
+                    int descriptionLength = reader.GetByte();
```
```suggestion
                    byte descriptionLength = reader.GetByte();
```

drewnoakes replied:
I'd use the actual type here rather that widening to a signed type. It'll prevent wondering whether it may be less than zero below.

kwhopper commented:
Removed the int cast of descriptionLength. It initially had to be int because the Java code adds pos to it before entering the loop so it was either start with an int or cast pos. I removed that logic and used a decrement of descriptionLength in the loop instead. Also fixed a regression when using the new clippingpath constants.

If this looks ok to you, I'll go ahead and merge. Thanks!

drewnoakes commented:
Thanks again.
